pkg/volume/photon_pd/attacher.go

```diff
@@ -0,0 +1,295 @@
+}
+
+// MountDevice mounts device to global mount point.
+func (attacher *photonPersistentDiskAttacher) MountDevice(spec *volume.Spec, devicePath string, deviceMountPath string) error {
+	mounter := attacher.host.GetMounter()
+	notMnt, err := mounter.IsLikelyNotMountPoint(deviceMountPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			if err := os.MkdirAll(deviceMountPath, 0750); err != nil {
+				glog.Errorf("Failed to create directory at %#v. err: %s", deviceMountPath, err)
+				return err
+			}
+			notMnt = true
+		} else {
+			return err
+		}
+	}
+
+	volumeSource, _, err := getVolumeSource(spec)
+	if err != nil {
+		glog.Errorf("Photon Controller attacher: MountDevice failed to get volume source. err: %s", err)
+		return err
+	}
+
+	options := []string{}
+
+	if notMnt {
+		diskMounter := &mount.SafeFormatAndMount{Interface: mounter, Runner: exec.New()}
+		err = diskMounter.FormatAndMount(devicePath, deviceMountPath, volumeSource.FSType, options)
+		if err != nil {
+			os.Remove(deviceMountPath)
+			return err
+		}
+		glog.V(4).Infof("formatting spec %v devicePath %v deviceMountPath %v fs %v with options %+v", spec.Name(), devicePath, deviceMountPath, volumeSource.FSType, options)
+	}
+	return nil
+}
+
+type photonPersistentDiskDetacher struct {
+	mounter     mount.Interface
+	photonDisks photon.Disks
+}
+
+var _ volume.Detacher = &photonPersistentDiskDetacher{}
+
+func (plugin *photonPersistentDiskPlugin) NewDetacher() (volume.Detacher, error) {
+	photonCloud, err := getCloudProvider(plugin.host.GetCloudProvider())
+	if err != nil {
+		glog.Errorf("Photon Controller attacher: NewDetacher failed to get cloud provider. err: %s", err)
+		return nil, err
+	}
+
+	return &photonPersistentDiskDetacher{
+		mounter:     plugin.host.GetMounter(),
+		photonDisks: photonCloud,
+	}, nil
+}
+
+// Detach the given device from the given host.
+func (detacher *photonPersistentDiskDetacher) Detach(deviceMountPath string, nodeName types.NodeName) error {
+
+	hostName := string(nodeName)
+	pdID := deviceMountPath
+	attached, err := detacher.photonDisks.DiskIsAttached(pdID, nodeName)
+	if err != nil {
+		// Log error and continue with detach
+		glog.Errorf(
+			"Error checking if persistent disk (%q) is already attached to current node (%q). Will continue and try detach anyway. err=%v",
+			pdID, hostName, err)
+	}
+
+	if err == nil && !attached {
+		// Volume is already detached from node.
+		glog.V(4).Infof("detach operation was successful. persistent disk %q is already detached from node %q.", pdID, hostName)
+		return nil
+	}
+
+	attachdetachMutex.LockKey(hostName)
+	defer attachdetachMutex.UnlockKey(hostName)
+	if err := detacher.photonDisks.DetachDisk(pdID, nodeName); err != nil {
+		glog.Errorf("Error detaching volume %q: %v", pdID, err)
+		return err
+	}
+	return nil
+}
+
+func (detacher *photonPersistentDiskDetacher) WaitForDetach(devicePath string, timeout time.Duration) error {
+	ticker := time.NewTicker(checkSleepDuration)
```

Comment:
You should really just use `util.Poll` here.

https://github.com/kubernetes/kubernetes/blob/master/pkg/util/wait/wait.go#L161